dbms/src/Flash/Mpp/ExchangeReceiverCommon.h

```diff
@@ -0,0 +1,355 @@
+// Copyright 2022 PingCAP, Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <Common/FailPoint.h>
+#include <Common/MPMCQueue.h>
+#include <Common/MemoryTracker.h>
+#include <Common/ThreadManager.h>
+#include <Flash/Coprocessor/CHBlockChunkCodec.h>
+#include <Flash/Coprocessor/ChunkCodec.h>
+#include <Flash/Coprocessor/DecodeDetail.h>
+#include <Flash/Coprocessor/FineGrainedShuffle.h>
+#include <Flash/Mpp/TrackedMppDataPacket.h>
+#include <kvproto/mpp.pb.h>
+#include <tipb/executor.pb.h>
+#include <tipb/select.pb.h>
+
+namespace DB
+{
+namespace FailPoints
+{
+extern const char random_receiver_sync_msg_push_failure_failpoint[];
+extern const char random_receiver_async_msg_push_failure_failpoint[];
+} // namespace FailPoints
+
+constexpr Int32 batch_packet_count = 16;
+
+struct ReceivedMessage
+{
+    size_t source_index;
+    String req_info;
+    // shared_ptr<const MPPDataPacket> is copied to make sure error_ptr, resp_ptr and chunks are valid.
+    const std::shared_ptr<DB::TrackedMppDataPacket> packet;
+    const mpp::Error * error_ptr;
+    const String * resp_ptr;
+    std::vector<const String *> chunks;
+    bool is_local_tunnel_msg;
+
+    // Constructor that move chunks.
+    ReceivedMessage(size_t source_index_,
+                    const String & req_info_,
+                    const std::shared_ptr<DB::TrackedMppDataPacket> & packet_,
+                    const mpp::Error * error_ptr_,
+                    const String * resp_ptr_,
+                    std::vector<const String *> && chunks_,
+                    bool is_local_tunnel_msg_ = false)
+        : source_index(source_index_)
+        , req_info(req_info_)
+        , packet(packet_)
+        , error_ptr(error_ptr_)
+        , resp_ptr(resp_ptr_)
+        , chunks(chunks_)
+        , is_local_tunnel_msg(is_local_tunnel_msg_)
+    {}
+
+    void switchMemTracker()
+    {
+        if (is_local_tunnel_msg)
```

Comment:
We can remove this line, because https://github.com/pingcap/tiflash/blob/16cf52428256d51b9a37873e83381f85c5bb66f1/dbms/src/Flash/Mpp/TrackedMppDataPacket.h#L89.

Reply:
> We can remove this line, because of
> 
> https://github.com/pingcap/tiflash/blob/16cf52428256d51b9a37873e83381f85c5bb66f1/dbms/src/Flash/Mpp/TrackedMppDataPacket.h#L89
> 
> .

okk